cedar-lean/Cedar/Thm/Partial/Evaluation/WellFormed.lean

```diff
@@ -28,51 +28,53 @@ import Cedar.Thm.Data.Set
 
 namespace Cedar.Spec
 
-/--
-  We define `WellFormed` for `Value` in the obvious way
--/
-def Value.WellFormed (v : Value) : Prop :=
-  match v with
-  | .set s => s.WellFormed
-  | .record r => r.WellFormed
+open Cedar.Data
+
+/-- All `Prim`s are structurally WellFormed. -/
+def Prim.WellFormed : Spec.Prim → Prop
   | _ => true
 
-/--
-  `Request`s are `WellFormed` if the context is `WellFormed`
--/
-def Request.WellFormed (req : Request) : Prop :=
-  req.context.WellFormed
+/-- All `Ext`s are structurally WellFormed. -/
+def Ext.WellFormed : Ext → Prop
+  | .decimal _ => true
+  | .ipaddr _  => true
+
+def Value.WellFormed : Spec.Value → Prop
```

Comment:
Not a blocker to merging the PR, but you may find it easier to do proofs over an inductive predicate that defines well-formedness rather than an explicit recursive function. Something like [this](https://github.com/cedar-policy/cedar-spec/blob/57f3da8af7c42c16529934f601f72fbf7dba6eb3/cedar-lean/Cedar/Thm/Validation/Typechecker/Types.lean#L49).

The inductive predicate approach doesn't require explicit termination arguments in the definition, and you can do induction directly on an assumption that says an argument is well-formed.

Reply:
I'm trying this, but finding it less convenient for proofs about well-formedness (could be user error or inexperience).  For instance, to discharge an obligation `Partial.Value.WellFormed (Partial.Value.value (Spec.Value.record m))`, previously I could use `simp [Partial.Value.WellFormed, Spec.Value.WellFormed]` to convert this to a pair of obligations `m.WellFormed` and `∀ v, ...`.  With the `inductive` definition of `Partial.Value.WellFormed` and `Spec.Value.WellFormed`, instead of one `simp` I seem to now need two `apply`s to accomplish the same thing: `apply Partial.Value.WellFormed.wf_val (.record m)` followed by `apply Spec.Value.WellFormed.wf_record m`.

Reply:
> I seem to now need two `apply`s to accomplish the same thing: `apply Partial.Value.WellFormed.wf_val (.record m)` followed by `apply Spec.Value.WellFormed.wf_record m`.

I like to use one `exact` for these; something like `exact Partial.Value.WellFormed.wf_val (.record m) (Spec.Value.WellFormed.wf_record m)`.

Reply:
Has to be one `apply` in my case (as I want to leave one obligation as a goal) and it's a lot uglier than the original `simp`, but maybe it's more precise / less fragile in some sense.

Reply:
I'm more concerned about the second problem.  One thing that is fundamentally different about the `inductive` definition is that we could have multiple arms of `Partial.Value.WellFormed` that apply to the same arm of `Partial.Value`.  That generality is powerful, but not needed here, and one negative consequence of it is that when Lean sees `Partial.Value.WellFormed (Partial.Value.value v)` it doesn't immediately know which arm of `Partial.Value.WellFormed` to use, as it does with the `def`.  That prevents the `simp` from working.  I don't know how to tell Lean to use a specific arm of `Partial.Value.WellFormed` inside the consequent of a hypothesis like this.

Reply:
> Later in the same proof, I have a hypothesis `Partial.Value.value v ∈ Map.values request.context → Partial.Value.WellFormed (Partial.Value.value v)`. Previously, I could use `simp [Partial.Value.WellFormed] at h` to simplify this to `Partial.Value.value v ∈ Map.values request.context → Spec.Value.WellFormed v`. It's unclear how I can accomplish the same transformation with `Partial.Value.WellFormed` defined as an `inductive`. By which I mean, probably there is a way but I don't know the right tactic?

I haven't looked at that particular proof in detail, but for hypotheses that look like this, I get rid of the implication first by applying `h` to the antecedent.  When left with `h : Partial.Value.WellFormed (Partial.Value.value v)`.  Then, if you do `cases h`, Lean will deconstruct `h` similarly to how it deconstructs an inductive data type.

Alternatively, if you prefer to keep the implication, you can a prove lemma that says `Partial.Value.WellFormed (Partial.Value.value v) → Spec.Value.WellFormed v` and use those with `rw`.  These proofs showing what's implied by an inductive predicate are quick to do with `cases`.

In general, inductive predicates are very nice to work with if you need to perform induction on a value/expr/etc that's WellFormed.  See, e.g., [here](https://github.com/leanprover-community/batteries/blob/60d622c124cebcecc000853cdae93f4251f4beb5/Batteries/Data/RBMap/WF.lean#L507-L511).  Another benefit is that they are idiomatic :)  For example, the standard library uses them consistently for predicate definitions over inductive types (e.g., [RBNode.WF](https://github.com/leanprover-community/batteries/blob/60d622c124cebcecc000853cdae93f4251f4beb5/Batteries/Data/RBMap/Basic.lean#L602-L611), [List.Forall_2](https://leanprover-community.github.io/mathlib4_docs/Batteries/Data/List/Basic.html#List.Forall%E2%82%82), [List.perm](https://github.com/leanprover-community/batteries/blob/60d622c124cebcecc000853cdae93f4251f4beb5/Batteries/Data/List/Basic.lean#L1349-L1357), etc).

Again, not a merge blocker by any means.  We can change this later if we want to.

Reply:
In this case I need to keep the implication because my goal is `Spec.Value.WellFormed v` and I want to apply this hypothesis.

Noted about using `rw` and a lemma.  And noted about not being a merge blocker.  I'll work on this for a subsequent PR.